### PR TITLE
Improvements to sensitivity computation

### PR DESCRIPTION
**Pull request by rsanfer**
This PR incorporates the computation of the sensitivity with respect to the forces applied to the beam.
To compile, use `make reverse` in the code folder, and run the script runbeam.py.

### CODE REVIEW

File: pyBeam/Makefile [hunk 1]
```diff
@@ -6,15 +6,52 @@ CODI_PATH = /home/rsanfer/pyBeam/pyBeam/CoDiPack/include
...
+       ./src/Rotations.cpp \
+       ./src/beam.cpp
 
-all:
-	swig -c++ -python -Wall -py3 pyBeam.i
-	# I need to add all the requested .cpp files used in the DLM.cpp for the relative .o files... -DDEBG
-	g++ -O2 -c -w -std=gnu++11 -DLINUX=1 -fPIC pyBeam_wrap.cxx ./src/ROMFM_FULLORDER.cpp ./src/input.cpp ./src/FiniteElement.cpp ./src/StructSyst.cpp ./src/Rotations.cpp ./src/beam.cpp -I/usr/include/python3.7m $(INCLPATH)
-	# ... That need to be used here so that the relative pyDLM_Cpp_wrap.o compiles the correct _pyDLM_Cpp.so
-	g++ -O2 -shared -fPIC ROMFM_FULLORDER.o input.o FiniteElement.o StructSyst.o Rotations.o beam.o pyBeam_wrap.o -o _pyBeam.so  $(INCLPATH) 
+_OBJS = ./obj/ROMFM_FULLORDER.o \
```
Comment (rsanfer) [inline pyBeam/Makefile:18]: This makes the makefile more readable

File: pyBeam/include/beam.h [hunk 2]
```diff
@@ -32,41 +32,65 @@
...
   void Solve(void);
 
-  passivedouble ExtractDisplacements(int iNode, int iDim);
+  passivedouble OF_NodeDisplacement(int iNode);
+
+  passivedouble ComputeAdjoint(void);
+
+  // Inlined functions
+
+  inline void SetThickness(passivedouble val_thickness) {thickness = val_thickness;}
```
Comment (rsanfer) [inline pyBeam/include/beam.h:80]: I have inlined the functions for readability and efficiency.

File: pyBeam/include/datatypes/ad_forward.hpp [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+ * General Public License along with pyBeam.
+ * If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+#pragma once
+
+#include "../../CoDiPack/include/codi.hpp"
+
+typedef codi::RealForward addouble;
```
Comment (rsanfer) [inline pyBeam/include/datatypes/ad_forward.hpp:31]: This is the datatype for the forward mode.

File: pyBeam/include/datatypes/ad_passive.hpp [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+ * See the GNU Affero General Public License for more details.
+ * You should have received a copy of the GNU Affero
+ * General Public License along with pyBeam.
+ * If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+#pragma once
+
+typedef double addouble;
```
Comment (rsanfer) [inline pyBeam/include/datatypes/ad_passive.hpp:29]: This is the passive double datatype.

File: pyBeam/include/datatypes/ad_reverse.hpp [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+ * General Public License along with pyBeam.
+ * If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+#pragma once
+
+#include "../../CoDiPack/include/codi.hpp"
+
+typedef codi::RealReverse addouble;
```
Comment (rsanfer) [inline pyBeam/include/datatypes/ad_reverse.hpp:31]: This is the reverse mode datatype.

File: pyBeam/include/geometry.h [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+class CNode {
+private:
+
+protected:
+
+  addouble *coord;
+
+public:
+
+  CNode(void);
```
Comment (rsanfer) [inline pyBeam/include/geometry.h:41]: I added the CNode class so we can store the geometrical features of each node.

File: pyBeam/include/geometry.h [hunk 7]
```diff
@@ -0,0 +1,69 @@
...
+class CConnectivity {
+private:
+
+protected:
+
+public:
+
+  CNode **node;
+
+  CConnectivity(void);
```
Comment (rsanfer) [inline pyBeam/include/geometry.h:61]: And this for the connectivity, to read from a mesh file.

File: pyBeam/src/ad.cpp [hunk 8]
```diff
@@ -0,0 +1,44 @@
...
+#include "../include/datatypes/ad_forward.hpp"
+#else
+#include "../include/datatypes/ad_passive.hpp"
+#endif
+
+namespace AD {
+#ifdef CODI_REVERSE_TYPE
+
+  /*--- Initialization of the tape ---*/
+  addouble::TapeType& globalTape = addouble::getGlobalTape();
```
Comment (rsanfer) [inline pyBeam/src/ad.cpp:41]: This initiates the tape.

File: pyBeam/src/beam.cpp [hunk 9]
```diff
@@ -203,57 +188,58 @@ void CBeamSolver::Solve(void){
...
-  //globalTape.evaluate();
-  
-  //grad_t = thickness.getGradient();
-   
-  //std::cout << " t' is " << grad_t << endl;
-
 }
 
-passivedouble CBeamSolver::ExtractDisplacements(int iNode, int iDim){
+passivedouble CBeamSolver::OF_NodeDisplacement(int iNode){
```
Comment (rsanfer) [inline pyBeam/src/beam.cpp:193]: This is the objective function, computes the norm of the displacement of node iNode.

**Review by rsanfer — COMMENTED** [review]
This is an example review. This way we can interact through here and see each other changes when we push them into develop.

File: pyBeam/Makefile (hunk 1 above)
Comment (RoccoBombardieri) [reply pyBeam/Makefile:18]: Great,

I would add a further option (with time) addressing the python specific version we use (if 2.7 or 3.7, which is the one you are using I guess). 

This is necessary in lines:
swig -c++ -python -Wall **-py3** ./swig/pyBeam.i
g++ -O2 -c -w -std=gnu++11 -DLINUX=1 -fPIC ./swig/pyBeam_wrap.cxx $(_SRC) **-I/usr/include/python3.7m** $(INCLPATH)$(INCLPATH)

File: pyBeam/include/geometry.h (hunk 6 above)
Comment (RoccoBombardieri) [reply pyBeam/include/geometry.h:41]: Good, If the storing type is the SU2 one I will refer to that for my part, I'll give it a look soon

**Review by RoccoBombardieri — APPROVED** [review]
Ok, I'm not deep enough into some parts of the code to review all the changes but I checked everything. One note regarding the setloads. I have in mind a load file in the format: NODE_ID DOF_ID LOAD_MODULE developed on different lines. This may need a firther change in the py part of the code. At the moment the input reading is in the SU2 style (you can check it in the other branch)

File: pyBeam/Makefile (hunk 1 above)
Comment (rsanfer) [reply pyBeam/Makefile:18]: Ok, I'll keep that one in mind. I'll add it to the tab "Projects" (see above)

**Comment by rsanfer** [comment]
> Ok, I'm not deep enough into some parts of the code to review all the changes but I checked everything. One note regarding the setloads. I have in mind a load file in the format: NODE_ID DOF_ID LOAD_MODULE developed on different lines. This may need a firther change in the py part of the code. At the moment the input reading is in the SU2 style (you can check it in the other branch)

That we can always do, however the loads will come from the fluid part (so there is no need for a file there, the'll be passed by reference)

**Pull request merged**